### PR TITLE
Bws108 upload object to space

### PR DESCRIPTION
**Pull request by ania-leichenko**

### CODE REVIEW

File: backend/src/common/enums/exception/exception-message.enum.ts [hunk 1]
```diff
@@ -34,6 +34,7 @@ const ExceptionMessage = {
   NOTHING_TO_UPDATE: 'Nothing to update',
   WORKER_DELETE_WORKER: 'Worker is not able to delete another worker',
   SPACE_EXISTS: 'Space with this name already exists',
+  OBJECT_NOT_UPLOADED: 'Object not uploaded',
```
Comment (what1s1ove) [inline backend/src/common/enums/exception/exception-message.enum.ts]:
```suggestion
  OBJECT_NOT_UPLOADED: 'Object is not uploaded',
```

File: backend/src/common/types/bs-object/bs-object-upload-payload.type.ts [hunk 2]
```diff
@@ -0,0 +1,2 @@
+import { File } from 'fastify-multer/lib/interfaces';
+export { type File };
```
Comment (what1s1ove) [inline backend/src/common/types/bs-object/bs-object-upload-payload.type.ts]:
```suggestion
export { type File } from 'fastify-multer/lib/interfaces';
```

File: backend/src/exceptions/bs-object-error/bs-object-error.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { ExceptionMessage, HttpCode } from '~/common/enums/enums';
```
Comment (what1s1ove) [inline backend/src/exceptions/bs-object-error/bs-object-error.ts]:
```suggestion
```
we should have only `BsError`. remove this one

File: backend/src/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { File } from 'fastify-multer/lib/interfaces';
+
+declare module 'fastify' {
+  interface FastifyRequest {
+    file: File;
```
Comment (what1s1ove) [inline backend/src/index.d.ts]:
```suggestion
    file?: File;
```

File: backend/src/middlewares/middlewares.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export * from './upload/upload.middleware';
```
Comment (what1s1ove) [inline backend/src/middlewares/middlewares.ts]: avoid star import

File: shared/src/common/enums/api/bs-api-path.enum.ts [hunk 6]
```diff
@@ -1,6 +1,7 @@
 enum BSApiPath {
   ROOT = '/',
   SPACES = '/spaces',
+  SPACES_FILE = '/spaces/:id/objects',
```
Comment (what1s1ove) [inline shared/src/common/enums/api/bs-api-path.enum.ts]:
```suggestion
  SPACES = '/spaces',
```
create a separate enum as for `spaces`

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/api/bs/bs.api.ts [hunk 7]
```diff
@@ -88,6 +94,26 @@ const initBsApi: FastifyPluginAsync<Options> = async (fastify, opts) => {
       return rep.send(true).status(HttpCode.OK);
     },
   });
+
+  fastify.route<{
+    Params: ObjectUploadParamsDto;
+    File: File;
+  }>({
+    method: HttpMethod.POST,
+    url: BSApiPath.SPACES_FILE,
```
Comment (Donymak) [inline backend/src/api/bs/bs.api.ts]:
```suggestion
    url: BSObjectApiPath.SPACES_FILE,
```

File: backend/src/api/bs/bs.api.ts [hunk 8]
```diff
@@ -1,5 +1,8 @@
 import { FastifyPluginAsync, FastifyRequest } from 'fastify';
-import { space as spaceServ } from '~/services/services';
+import {
+  space as spaceServ,
+  bsObject as bsObjectServ,
+} from '~/services/services';
 import {
   HttpCode,
   HttpMethod,
```
Comment (Donymak) [inline backend/src/api/bs/bs.api.ts:8]:
```suggestion
  HttpMethod,
  BSObjectApiPath,
```

File: frontend/src/components/eam/eam.tsx [hunk 9]
```diff
@@ -39,6 +39,25 @@ const EAM: FC = () => {
...
+  const handleGroupsReload = (): void => {
+    dispatch(
+      eamActions.loadGroups({
+        tenantId: tenantId as string,
+        from: 0,
+        count: 5,
+      }),
+    );
+  };
+
```
Comment (Donymak) [inline frontend/src/components/eam/eam.tsx]: it should've been merged... 
IDK why is it green if it already exists on develop

File: backend/src/services/bs-object/bs-object.service.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    token,
+    file,
+    id,
+  }: UploadPayload): Promise<BSObjectEntity> {
+    const user: TokenPayload = await this.#tokenService.decode(token);
+
+    const space = await this.#spaceService.getSpacesById(id);
+
+    const worker = await this.#workerService.getUserById(space.createdBy);
+    const isCurrentWorker = worker?.user?.tenantId === user.tenantId;
```
Comment (Yaroslaww-1) [inline backend/src/services/bs-object/bs-object.service.ts]:
```suggestion
    const isCurrentWorkerOwnerOfSpace = worker?.user?.tenantId === user.tenantId;
```

File: backend/src/services/bs-object/bs-object.service.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        status: HttpCode.BAD_REQUEST,
+        message: ExceptionMessage.OBJECT_NOT_UPLOADED,
+      });
+    }
+    const objectUploadEntity = BSObjectEntity.createNew({
+      name: file.originalname,
+      sizeInBytes: file.size as number,
+      spaceId: space.id,
+      uploadedBy: user.userId,
+      awsObjectKey: awsObjectKey.substring(1, -1),
```
Comment (Yaroslaww-1) [inline backend/src/services/bs-object/bs-object.service.ts]: Are you sure that it is correct? Currently It will return only one char

**Review by Yaroslaww-1 — CHANGES_REQUESTED** [review]

File: backend/src/services/bs-object/bs-object.service.ts (hunk 11 above)
Comment (Donymak) [reply backend/src/services/bs-object/bs-object.service.ts]: AWS returns key as "key", so:
```suggestion
      awsObjectKey: awsObjectKey.slice(1, -1),
```

File: shared/src/common/enums/api/bs-object-path.enum.ts [hunk 12]
```diff
@@ -0,0 +1,5 @@
+enum BSObjectApiPath {
+  SPACES_FILE = '/spaces/:id/objects',
```
Comment (what1s1ove) [inline shared/src/common/enums/api/bs-object-path.enum.ts]: add this path to https://github.com/BinaryStudioAcademy/bsa-winter-2021-2022-bws/blob/d19b44ef6f5d0d6743296369369f1f02516ccabc/shared/src/common/enums/api/spaces-api-path.enum.ts

Comment (what1s1ove) [inline shared/src/common/enums/api/bs-object-path.enum.ts]:
```suggestion
  SPACES_$ID__FILES = '/spaces/:id/objects',
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by Donymak — DISMISSED** [review]

File: shared/src/common/enums/api/spaces-api-path.enum.ts [hunk 13]
```diff
@@ -1,6 +1,7 @@
 enum SpacesApiPath {
   ROOT = '/',
   $ID = '/:id',
+  SPACES_$ID__FILES = '/spaces/:id/objects',
```
Comment (what1s1ove) [inline shared/src/common/enums/api/spaces-api-path.enum.ts]:
```suggestion
  $ID_OBJECTS = '/:id/objects',
```

File: backend/src/api/bs/bs.api.ts [hunk 14]
```diff
@@ -88,6 +94,26 @@ const initBsApi: FastifyPluginAsync<Options> = async (fastify, opts) => {
       return rep.send(true).status(HttpCode.OK);
     },
   });
+
+  fastify.route<{
+    Params: ObjectUploadParamsDto;
+    File: File;
+  }>({
+    method: HttpMethod.POST,
+    url: SpacesApiPath.SPACES_$ID__FILES,
```
Comment (what1s1ove) [inline backend/src/api/bs/bs.api.ts]:
```suggestion
    url: `${BSApiPath.SPACES}${SpacesApiPath.$ID_OBJECTS}`,
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/services/space/space.service.ts [hunk 15]
```diff
@@ -103,6 +115,18 @@ class Space {
 
     await this.#spaceRepository.delete(id);
   }
+
+  public async getSpacesById(id: string): Promise<SpaceEntity> {
```
Comment (Yaroslaww-1) [inline backend/src/services/space/space.service.ts]:
```suggestion
  public async getSpaceById(id: string): Promise<SpaceEntity> {
```

**Review by Yaroslaww-1 — DISMISSED** [review]

**Review by what1s1ove — APPROVED** [review]

**Review by Yaroslaww-1 — APPROVED** [review]

**Pull request merged**